datastream-server/src/main/java/com/linkedin/datastream/server/zk/ZkAdapter.java

```diff
@@ -728,6 +731,73 @@ private void removeTaskNode(String instance, String name) {
     _zkclient.deleteRecursive(instancePath);
   }
 
+  /**
+   * Update the task assignment for the given instances and issue assignment tokens for stopping tasks.
+   * Assignment tokens are used by the follower hosts to signal the leader that they finished handling their assignments.
+   *
+   * In addition to the work done by updateAllAssignments, this method writes to
+   * <ol>
+   *    *  <li>{@code /<cluster>/dms/<datastream-name>/assignmentTokens/<token1>,<token2>...}</li>
+   * </ol>
+   */
+  public void updateAllAssignmentsAndIssueTokens(Map<String, List<DatastreamTask>> assignmentsByInstance,
+      List<DatastreamGroup> stoppingDatastreamGroups) {
+    Validate.notNull(assignmentsByInstance, "null assignmentsByInstance");
+    Validate.notNull(stoppingDatastreamGroups, "null stoppingDatastreamGroups");
+
+    Map<String, Set<DatastreamTask>> currentAssignment = getAllAssignedDatastreamTasks();
+    Set<String> stoppingDatastreamTaskPrefixes = stoppingDatastreamGroups.stream().
+        map(DatastreamGroup::getTaskPrefix).collect(toSet());
+    Map<String, DatastreamGroup> taskPrefixDatastreamGroups = stoppingDatastreamGroups.stream().
+        collect(Collectors.toMap(DatastreamGroup::getTaskPrefix, Function.identity()));
+
+    // For each stopping datastream group, find all the instances that currently have tasks assigned
+    Map<DatastreamGroup, Set<String>> stoppingDgInstances = new HashMap<>();
+    for (String instance : currentAssignment.keySet()) {
+        Set<DatastreamTask> instanceTasks = currentAssignment.get(instance);
+        List<DatastreamTask> stoppingTasks = instanceTasks.stream().
+            filter(t -> stoppingDatastreamTaskPrefixes.contains(t.getTaskPrefix())).
+            collect(Collectors.toList());
+        for (DatastreamTask stoppingTask : stoppingTasks) {
+          DatastreamGroup datastreamGroup = taskPrefixDatastreamGroups.get(stoppingTask.getTaskPrefix());
+          stoppingDgInstances.computeIfAbsent(datastreamGroup, k -> new HashSet<>()).add(instance);
+        }
+      }
+
+    String hostname = "localhost";
+    try {
+      hostname = InetAddress.getLocalHost().getHostName();
+    } catch (UnknownHostException ex) {
+      LOG.warn("Unable to obtain hostname for leader");
+    }
+
+    // Issue assignment tokens for each instance and each stopping datastream
+    for (DatastreamGroup stoppingGroup : stoppingDatastreamGroups) {
+      for (Datastream stoppingStream : stoppingGroup.getDatastreams()) {
+        String path = KeyBuilder.datastream(_cluster, stoppingStream.getName());
+
+        if (!_zkclient.exists(path)) {
+          LOG.warn("Trying to issue assignment tokens for non-existing stream: " + stoppingStream.getName());
```

Comment:
```suggestion
          LOG.warn("Trying to issue assignment tokens for non-existing stream={}", stoppingStream.getName());
```

Reply:
Your version is not necessarily better than string concatenation in mine. Concatenation is usually faster than formatting a string too. I prefer formatting when there are more variables involved and when you need finer control wrt how those variables are formatted in the final string.

Interesting SO post that I found:
https://stackoverflow.com/questions/925423/is-it-better-practice-to-use-string-format-over-string-concatenation-in-java

Reply:
I don't think the performance would be a problem here, I would go with consistency and maintain the logging format.

Reply:
We have mixed uses for logging API calls everywhere. And again, this is more of a personal preference.